fix: NetworkTransform authority warning [MTT-1291]

NoelStephensUnity opened the pull request:
Authority warning in NetworkTransform for rotation should be approximated. Quaternion to Euler conversion isn't deterministic and triggers a lot of false positives.

[MTT-1291](https://jira.unity3d.com/browse/MTT-1291)

## Changelog
- Fixed: issue where the NetworkTransform authority warning message could generate false positive results.

## Testing and Documentation
- Includes an integration test.

NoelStephensUnity committed:
fix

enabling the NetworkTransform debug information.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
test

This verifies the authority warning is working on the client side.

NoelStephensUnity committed:
style

MTT-1291
Added some additional comments and making sure this branch is tracked in the task.

NoelStephensUnity committed:
update

Added Unity.Netcode.Components to test project's runtime assembly so we can write tests for the components in this test assembly.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
style

white space removal

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -920,15 +919,14 @@ protected virtual void Update()
                         // there are several bugs in this code, as we the message is dumped out under odd circumstances
                         //  For Matt, it would trigger when an object's rotation was perturbed by colliding with another
```
Can probably remove my name at this point :)

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -920,15 +919,14 @@ protected virtual void Update()
                         //  For Matt, it would trigger when an object's rotation was perturbed by colliding with another
                         //  object vs. explicitly rotating it
-                        if (oldStateDirtyInfo.isPositionDirty || oldStateDirtyInfo.isScaleDirty || (oldStateDirtyInfo.isRotationDirty && SyncRotAngleX && SyncRotAngleY && SyncRotAngleZ))
+                        if (oldStateDirtyInfo.isPositionDirty || oldStateDirtyInfo.isScaleDirty || (oldStateDirtyInfo.isRotationDirty && (SyncRotAngleX || SyncRotAngleY || SyncRotAngleZ)))
```
Interesting.  This code is better, for sure.  However, we can just remove the check for SyncRotAngle{XYZ}, because the code that makes it dirty only does so when the corresponding SyncRotAngle{XYZ} is true.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
update

only need to check for isRotationDirty as this is only set when any of the SyncRotAngle[xyz] are set to true.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
update

Fixes the issue where there was no minimum threshold value which meant a user could set the threshold to zero which would cause non-authoritative false-positive warnings.
Fixes the issue where NetworkRigidBody was only setting IsKinematic to true when the NetworkObject was spawned which could cause non-authoritative false-positive warnings.
Fixes the issue where the very next frame update after OnNetworkSpawn could cause non-authoritative false-positive warnings.
Fixes the issue where changing the parent could cause non-authoritative false-positive warnings.
Fixes the issue where rotation was improperly checking for a delta rotation that was larger than the rotation threshold.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -5,50 +5,82 @@ namespace Unity.Netcode.Components
             if (m_IsAuthority == false)
             {
-                m_OriginalKinematic = m_Rigidbody.isKinematic;
+                //m_OriginalKinematic = m_Rigidbody.isKinematic;
```
Leftover comment

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -15,12 +15,42 @@ namespace Unity.Netcode.Components
+            // or equal to the ThresholdMinimum
+            if (PositionThreshold < ThresholdMinimum)
+            {
+                PositionThreshold = ThresholdMinimum;
```
I think we should warn the user that we are overriding their settings in a log.  Then again, is this necessary if you're guarding against this with your editor Min / Range setttings?

NoelStephensUnity replied:
Well, this was my current train of thought on this:
We have a pre-release that allows users to set their threshold values to 0, and if users currently have any NetworkPrefabs (prefabs) with NetworkTransform threshold values already serialized as 0 then without OnValidate a user will have to manually edit the values for them to be kept within the minimum range.  So, for pre-existing users this will assure that in the event they have a threshold value set to zero it will be adjusted to the minimum threshold range automatically when they enter into playmode or make a build.

0xFA11 replied:
+1 for warning logs to inform user here

NoelStephensUnity replied:
Added warning messages.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -738,8 +784,25 @@ private void Awake()
             }
         }
 
+        /// <summary>
```
Can you do a 'see also' in your XMLDoc vs. repeating this comment?

NoelStephensUnity committed:
style

removing commented out code

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -917,23 +991,20 @@ protected virtual void Update()
                         var oldStateDirtyInfo = ApplyTransformToNetworkStateWithInfo(ref m_PrevNetworkState, 0, m_Transform);
 
                         // there are several bugs in this code, as we the message is dumped out under odd circumstances
```
Is this comment still appropriate?  Did we not fix it here?

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -904,8 +979,7 @@ protected virtual void Update()
                 {
-#if NGO_TRANSFORM_DEBUG
-                    if (m_CachedNetworkManager.LogLevel == LogLevel.Developer)
+                    if (m_CachedNetworkManager.LogLevel == LogLevel.Developer && m_AuthorityCheckDelay < Time.realtimeSinceStartup)
```
We seem to be going to a lot of effort here with this delay all to hide the warning message.  What if we simply don't show it?  Is it something useful to users; something they can act on?  Can the test coverage be enough where we don't feel the need to emit it?

NoelStephensUnity replied:
If a user has a bug in their code where they are trying to make changes to the NetworkTransform when they don't have authority, then without this message they may never know that they are doing this or have no easy way to check to make sure they are not trying to change a transform without authority.

_It would be difficult to write a tests for the scenario where a user's code is potentially changing the position, rotation, or scale of a NetworkObject that does not have the authority to make the change.  This would be more of an "improper usage" scenario which is better to provide some tool/way to detect this._

NoelStephensUnity replied:
As a side note:
This PR really turned into more of a "fix several issues" in associated areas, but most importantly fixes the delta rotation not exceeding the minimum threshold.  Due to the way it was being done, we would be sending delta updates to rotation due to false-positive results because we were not finding the shortest Euler angle between local and state transform values but just the absolute value of the difference between the local and state transform values.

The additional stuff for the warnings only happens when LogLevel.Developer, so it shouldn't have any real impact on performance while also providing a way for users to determine if they are trying to change transform values without authority in their code base.

mattwalsh-unity replied:
I'm somewhat ok with telling users that they are writing to a transform that they shouldn't be writing to - if we can do it in a water-tight way.  I'm not thrilled about the 1-tick delay and associated code that goes with it.  Can we fundamentally explain why it's precisely 1 tick, or did we add a 1 tick delay and the problem went away?  If the latter, how do we know it's 1 and only 1 tick for all platforms / situations?

NoelStephensUnity replied:
I believe 1 tick is a gross overshoot of the required time.  Let me see if there is a better way to handle this.

mattwalsh-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -917,23 +991,20 @@ protected virtual void Update()
+                            // ignoring rotation dirty since quaternions will mess with Euler angles, making this impossible to determine if the change to a single axis comes
+                            // from an unauthorized transform change or Euler to quaternion conversion artifacts.
                             var dirtyField = oldStateDirtyInfo.isPositionDirty ? "position" : oldStateDirtyInfo.isRotationDirty ? "rotation" : "scale";
                             Debug.LogWarning($"A local change to {dirtyField} without authority detected, reverting back to latest interpolated network state!", this);
```
I'm starting to think we just ditch the message (and the delay complexity).  We seem to be going to a lot of trouble here to detect and rollback for the users that are using the API in the incorrect way.

I empathize that we want to tell the users that they have tried to move the transform the wrong way, but this is always going to be hard to reliably detect (with all the numeric issues that go along with it) because we are listening to the position / rotate changes without a way to get in front of those changes and error out if they're in the wrong mode.

NoelStephensUnity replied:
Just a few lines down there is this:
```
                    // Apply updated interpolated value
                    ApplyInterpolatedNetworkStateToTransform(m_ReplicatedNetworkState.Value, m_Transform);
```
This will automatically revert any m_Transform changes back to the m_ReplicatedNetworkState.Value.

The logging block only detects this is happening but does not actually make any changes.
When LogLevel is not Developer, it will still do this (revert back) without the error messages.  
The reasons users might want to know this is happening is that it is "wasted" processing that is happening on all NetworkObject instances that don't have authority, so this is a check that users could also do during their optimization phase.

NoelStephensUnity committed:
style

updating an removing some comments.

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -5,50 +5,81 @@ namespace Unity.Netcode.Components
-        /// </summary>
-        private bool HasAuthority => m_NetworkTransform.CanCommitToTransform;
+        // Used to cache the authority state of this RigidBody during the last frame
+        protected bool m_IsAuthority;
```
is this being exposed to public API intentionally? if so, why?

NoelStephensUnity replied:
Good catch and it will be reverted back in next commit.

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -5,50 +5,81 @@ namespace Unity.Netcode.Components
         }
 
-        // Puts the rigidbody in a kinematic non-interpolated mode on everyone but the server.
+        // Puts the RigidBody in a kinematic non-interpolated mode on everyone but the server.
```
```suggestion
        // Puts the Rigidbody in a kinematic non-interpolated mode on everyone but the server.
```

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -57,7 +88,7 @@ private void UpdateRigidbodyKinematicMode()
             else
             {
-                // Resets the rigidbody back to it's non replication only state. Happens on shutdown and when authority is lost
+                // Resets the RigidBody back to it's non replication only state. Happens on shutdown and when authority is lost
```
```suggestion
                // Resets the Rigidbody back to its non-replication only state. Happens on shutdown and when authority is lost
```

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -450,23 +490,23 @@ internal bool ApplyTransformToNetworkState(ref NetworkTransformState networkStat
 
             if (SyncRotAngleX &&
-                Mathf.Abs(networkState.RotAngleX - rotAngles.x) > RotAngleThreshold)
+                Mathf.Abs(Mathf.DeltaAngle(networkState.RotAngleX, rotAngles.x)) > RotAngleThreshold)
```
love these <3

0xFA11 commented on testproject/Assets/Tests/Runtime/NetworkTransformAuthorityTest.cs:
```diff
@@ -0,0 +1,73 @@
+            m_TestPrefab = CreateNetworkObjectPrefab("AuthorityTest");
+            var rigidBody = m_TestPrefab.AddComponent<Rigidbody>();
+            // Just to keep the test prefab instances from falling
+            rigidBody.useGravity = false;
```
lol :D

0xFA11 commented on testproject/Assets/Tests/Runtime/NetworkTransformAuthorityTest.cs:
```diff
@@ -0,0 +1,73 @@
+
+            // Modify the client-side instance's transform position to generate the warning
+            m_ClientNetworkObjectBehaviour.transform.position += Vector3.one;
+            LogAssert.Expect(LogType.Warning, $"A local change to position without authority detected, reverting back to latest interpolated network state!");
```
not for this PR but I started to think that we should have our log strings stored in a const variable so that we could pay no runtime cost but also have the exact string shared between places without breaking them accidentally.
again, not for this PR, this is more like a note to myself.

NoelStephensUnity replied:
Yeah, I actually was thinking we needed a better way to store all of this stuff in some form of asset (i.e. ScriptableObject or the like).  It would allow us to more easily modify messages and provide a way to more easily translate messages to a different language (localize) in the event that is needed.

0xFA11 reviewed (approved):
left some minor comments/suggestions here and there but overall looks good to me!

NoelStephensUnity committed:
update

Fixing some issues Fatih caught.
Refactoring when we have to delay for the authority check.  It is only needed when the parent changes.
Removing data size spam messages when LogLevel.Developer.
Adding scale and rotation to the test as well as verification that there is no notification when below the threshold.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
style

NoelStephensUnity committed:
update

Adding warning message when any threshold is below the minimum threshold value during OnValidate.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1291-networktransform-authority-warning

NoelStephensUnity committed:
fix

fixing some merge issues.

NoelStephensUnity committed:
update

removing authority warning related code.

NoelStephensUnity committed:
update

removing components reference since it is no longer needed.

NoelStephensUnity committed:
update

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Update CHANGELOG.md
